core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java

```diff
@@ -0,0 +1,479 @@
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.ResolvingFileIO;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.rest.requests.CreateNamespaceRequest;
+import org.apache.iceberg.rest.requests.CreateTableRequest;
+import org.apache.iceberg.rest.requests.UpdateNamespacePropertiesRequest;
+import org.apache.iceberg.rest.responses.ConfigResponse;
+import org.apache.iceberg.rest.responses.CreateNamespaceResponse;
+import org.apache.iceberg.rest.responses.GetNamespaceResponse;
+import org.apache.iceberg.rest.responses.ListNamespacesResponse;
+import org.apache.iceberg.rest.responses.ListTablesResponse;
+import org.apache.iceberg.rest.responses.LoadTableResponse;
+import org.apache.iceberg.rest.responses.UpdateNamespacePropertiesResponse;
+import org.apache.iceberg.util.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RESTSessionCatalog extends BaseSessionCatalog implements Configurable<Configuration>, Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(RESTSessionCatalog.class);
+  private final Function<Map<String, String>, RESTClient> clientBuilder;
+  private final Cache<String, Map<String, String>> headers = Caffeine.newBuilder().build();
+  private Map<String, String> baseHeaders = ImmutableMap.of();
+  private RESTClient client = null;
+  private String catalogName = null;
+  private Map<String, String> properties = null;
+  private ResourcePaths paths = null;
+  private Object conf = null;
+  private FileIO io = null;
+
+  public RESTSessionCatalog() {
+    this(new HTTPClientFactory());
+  }
+
+  RESTSessionCatalog(Function<Map<String, String>, RESTClient> clientBuilder) {
+    this.clientBuilder = clientBuilder;
+  }
+
+  @Override
+  public void initialize(String name, Map<String, String> props) {
+    Preconditions.checkArgument(props != null, "Invalid configuration: null");
+    this.baseHeaders = RESTUtil.extractPrefixMap(props, "header.");
+    ConfigResponse config = fetchConfig(props);
+    Map<String, String> mergedProps = config.merge(props);
+    this.client = clientBuilder.apply(mergedProps);
+    this.catalogName = name;
+    this.properties = mergedProps;
+    this.paths = ResourcePaths.forCatalogProperties(properties);
+    String ioImpl = properties.get(CatalogProperties.FILE_IO_IMPL);
+    this.io = CatalogUtil.loadFileIO(ioImpl != null ? ioImpl : ResolvingFileIO.class.getName(), properties, conf);
+  }
+
+  Map<String, String> headers(SessionContext context) {
+    return headers.get(context.sessionId(), id -> {
+      ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+      builder.putAll(baseHeaders);
+
+      if (context.identity() != null) {
```

Comment:
I notice here that although we use session Id as the key for the headers map, what we are setting are just headers for identity and credentials. Basically the identity and credentials are the determining factor of the cache content (`Catalog` in the `BaseSessionCatalog`, and header map here). This is very similar to https://github.com/prestodb/presto/blob/master/presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergResourceFactory.java#L88.

That means if those information do not change, the same cached content could be shared across sessions. For example, here if 2 sessions have the same identity and credential, they actually have the same headers. So does it make more sense to use `content.identity() + content.credential()` as the cache key? This could greatly improve the cache hit rate, instead of creating something new for each session. This is especially helpful if the session is short-lived, but the caller credential is shared across sessions.

Reply:
Yes, I agree that this is probably going to change. I'm not sure if `identity` and `credential` are the right things to track here. And in Trino we would probably want to set session ID to one or both of them to have sessions that are not simply single queries. I'm definitely open to suggestions!

My main concern right now was to get people looking at the new catalog interface and make sure we have consensus that it's a good idea. What metadata we track in a session is important, but secondary if we don't end up going this direction.

Reply:
I see. If we are focusing on the consensus part, I think it is pretty aligned with the [TrinoCatalog](https://github.com/trinodb/trino/blob/master/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/TrinoCatalog.java) interface, there is a converter needed for a Trino [ConnectorSession](https://github.com/trinodb/trino/blob/master/core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorSession.java) but that is overall straightforward.

The reason I bring up this identity and credential part here is that maybe Iceberg's `SessionContext` can be an interface but not an implementation, and we have `SessionCatalog<T extends SessionContext>`, and `SessionContext` has one important method `cacheKey()` that can be used to define if 2 sessions should use the same underlying catalog, and the cache key can on the extreme end be the session ID to have 1 different catalog for each session, or be a composite of identity and credential values.

In Trino, then we can have `IcebergConnectorSession implements SessionContext` and `TrinoCatalog implements SessionCatalog<IcebergConnectorSession>` as a way to simplify the `TrinoCatalog` class.